File: apps/menu-bar/macos/ExpoMenuBar-macOS/MenuBarModule.m
#import "MenuBarModule.h"
#import "AppDelegate.h"
#import <sys/utsname.h>

#import <React/RCTLog.h>
#import <ServiceManagement/ServiceManagement.h>

@implementation MenuBarModule
{
  bool hasListeners;
}

RCT_EXPORT_MODULE();


// Will be called when this module's first listener is added.
-(void)startObserving {
    hasListeners = YES;
    // Set up any upstream listeners or background tasks as necessary
}

// Will be called when this module's last listener is removed, or on dealloc.
-(void)stopObserving {
    hasListeners = NO;
    // Remove upstream listeners, stop unnecessary background tasks
}

- (NSArray<NSString *> *)supportedEvents {
  return @[@"onNewCommandLine", @"onCLIOutput"];
}

+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

- (NSDictionary *)constantsToExport
{
  return @{
    @"appVersion"         : [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"] ?: [NSNull null],
    @"buildVersion"       : [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"] ?: [NSNull null],
    @"initialScreenSize"  : @{
      @"height":  @([[NSScreen mainScreen] frame].size.height),
      @"width":  @([[NSScreen mainScreen] frame].size.width)
    },
  };
}

RCT_EXPORT_METHOD(exitApp)
{
  exit(0);
}

RCT_EXPORT_METHOD(runCommand:(NSString *)command
                  arguments:(NSArray *)arguments
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  NSTask *task = [[NSTask alloc] init];
  NSPipe *pipe = [NSPipe pipe];

  [task setLaunchPath:@"/bin/sh"];
  [task setArguments:@[@"-l", @"-c", [NSString stringWithFormat:@"PATH=\"$PATH:/usr/local/bin\"; %@ %@ ", command, [arguments componentsJoinedByString:@" "]]]];

  [task setStandardOutput:pipe];
  [task setStandardError:pipe];

  NSFileHandle *file = [pipe fileHandleForReading];
   __block NSString *returnOutput = @"";
  [task launch];

  NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
  [notificationCenter addObserverForName:NSFileHandleReadCompletionNotification
                                  object:file
                                    queue:nil
                              usingBlock:^(NSNotification *notification) {
                                  NSData *chunk = notification.userInfo[NSFileHandleNotificationDataItem];
                                  NSString *wholeOutput = [[NSString alloc] initWithData:chunk encoding:NSUTF8StringEncoding];
                                  NSArray *outputs = [wholeOutput componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];

                                  for (NSString *output in outputs) {
                                    if ([output isEqualToString:@""]) {
                                      continue;
                                    }

                                    returnOutput = [returnOutput stringByAppendingString:output];

                                    if(self->hasListeners && output.length > 0 && ![output isEqualToString:@"\n"]){
                                      [self sendEventWithName:@"onNewCommandLine" body:output];
                                    }
                                  }

                                  [file readInBackgroundAndNotify];
                              }];
  [notificationCenter addObserverForName:NSTaskDidTerminateNotification
                                  object:task
                                    queue:nil
                              usingBlock:^(NSNotification *notification) {
                                  [notificationCenter removeObserver:self];
                              }];

  [file readInBackgroundAndNotify];
  [task waitUntilExit];

  resolve(returnOutput);
}

NSString *getMachineHardwareName(void) {
    struct utsname sysinfo;
    int retVal = uname(&sysinfo);
    if (EXIT_SUCCESS != retVal) return nil;

    return [NSString stringWithUTF8String:sysinfo.machine];
}

NSString *getCliResourceNameForArch(void) {
    NSString *arch = getMachineHardwareName();
    if([arch isEqualToString:@"arm64"]){
      return @"orbit-cli-arm64";
    }

    return @"orbit-cli-x64";
}

RCT_EXPORT_METHOD(runCli:(NSString *)command
                  arguments:(NSArray *)arguments
                  listenerId:(nonnull NSNumber *)listenerId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    NSTask *task = [[NSTask alloc] init];
    NSPipe *pipe = [NSPipe pipe];

    NSString *executablePath = [[NSBundle mainBundle] pathForResource:getCliResourceNameForArch() ofType:nil];

    [task setLaunchPath:executablePath];
    [task setArguments:[@[command] arrayByAddingObjectsFromArray:arguments]];

    NSMutableDictionary *environment = [NSMutableDictionary dictionaryWithDictionary:[[NSProcessInfo processInfo] environment]];
    [environment addEntriesFromDictionary:@{@"EXPO_MENU_BAR": @YES}];

    // Retrieve the envVars from NSUserDefaults
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    NSDictionary *envVars = [userDefaults objectForKey:@"envVars"];

    // Check if the retrieved object is indeed a NSDictionary
    if ([envVars isKindOfClass:[NSDictionary class]] && [envVars count] > 0) {
      [environment addEntriesFromDictionary:envVars];
    }

    [task setEnvironment:environment];
    [task setStandardOutput:pipe];
    [task setStandardError:pipe];

    NSFileHandle *file = [pipe fileHandleForReading];
    NSMutableString *returnOutput = [NSMutableString string];
    __block BOOL hasReachedReturnOutput = NO;
    __block BOOL hasReachedError = NO;

    [file setReadabilityHandler:^(NSFileHandle * _Nonnull handle) {
      NSData *availableData = [handle availableData];
      NSString *wholeOutput = [[NSString alloc] initWithData:availableData encoding:NSUTF8StringEncoding];
      NSArray *outputs = [wholeOutput componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];

      for (NSString *output in outputs) {
        if ([output isEqualToString:@""]) {
          continue;
        }

        if(hasReachedReturnOutput || hasReachedError){
          [returnOutput appendString:output];
        } else if([output isEqualToString:@"---- return output ----"]) {
          hasReachedReturnOutput = YES;
        } else if([output isEqualToString:@"---- thrown error ----"]) {
          hasReachedError = YES;
        } else if(self->hasListeners && output.length > 0 && ![output isEqualToString:@"\n"]) {
          NSDictionary *eventData = @{
            @"listenerId": listenerId,
            @"output": output
          };
          [self sendEventWithName:@"onCLIOutput" body:eventData];
        }
      }
    }];

    [task setTerminationHandler:^(NSTask * _Nonnull task) {
      [file setReadabilityHandler:nil];

      if(hasReachedError){
        reject(@"CLI_ERROR", returnOutput, nil);
      } else {
        resolve(hasReachedReturnOutput ? returnOutput : nil);
      }
    }];

    [task launch];
  });
}

RCT_EXPORT_METHOD(setLoginItemEnabled:(BOOL)enabled
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  if (SMLoginItemSetEnabled((__bridge CFStringRef)@"dev.expo.AutoLauncher", enabled)) {
    resolve(nil);
  } else {
    reject(@"AUTO_LAUNCHER_ERROR", @"Failed to update login item status.", nil);
  }
}

RCT_EXPORT_METHOD(showMultiOptionAlert:(NSString *)title
                  message:(NSString *)message
                  options:(NSArray<NSString *> *)options
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  dispatch_async(dispatch_get_main_queue(), ^{
    NSAlert *alert = [[NSAlert alloc] init];
    [alert setMessageText:title];
    [alert setInformativeText:message];
    [alert addButtonWithTitle:@"OK"];
    [alert addButtonWithTitle:@"Cancel"];

    NSPopUpButton *popUpButton = [[NSPopUpButton alloc] initWithFrame:NSMakeRect(0, 0, 250, 24) pullsDown:NO];
    [popUpButton addItemsWithTitles:options];

    [alert setAccessoryView:popUpButton];

    NSInteger response = [alert runModal];
    if (response == NSAlertFirstButtonReturn) {
      resolve(@([popUpButton indexOfSelectedItem]));
    } else {
      reject(@"MULTI_OPTION_ALERT", @"Selection was canceled.", nil);
    }
  });
}


RCT_EXPORT_METHOD(openSystemSettingsLoginItems)
{
  if (@available(macOS 13.0, *)) {
    [SMAppService openSystemSettingsLoginItems];
  } else {
    NSString *url = @"x-apple.systempreferences:com.apple.LoginItems-Settings.extension";
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:url]];
  }
}

RCT_EXPORT_METHOD(setEnvVars:(NSDictionary *)envVars)
{
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  [userDefaults setObject:envVars forKey:@"envVars"];
  [userDefaults synchronize];
}

RCT_EXPORT_METHOD(openPopover)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [((AppDelegate *)[NSApp delegate]) openPopover];
  });
}

RCT_EXPORT_METHOD(closePopover)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [((AppDelegate *)[NSApp delegate]) closePopover];
  });
}

@end

File: apps/menu-bar/modules/menu-bar/ios/MenuBarModule.swift
import ExpoModulesCore
import ServiceManagement

private let onNewCommandLine = "onNewCommandLine"
private let onCLIOutput = "onCLIOutput"

public class MenuBarModule: Module {
  private var hasListeners = false
  
  public func definition() -> ModuleDefinition {
    Name("MenuBar")
    
    Events(onNewCommandLine, onCLIOutput)

    Constants([
      "appVersion": self.appContext?.constants?.constants()["nativeAppVersion"],
      "buildVersion": self.appContext?.constants?.buildVersion(),
      "initialScreenSize": [
        "width": NSScreen.main?.frame.width,
        "height": NSScreen.main?.frame.height
      ],
      "homedir": NSHomeDirectory(),
    ])
    
    Function("exitApp") {
      exit(0)
    }
    
    AsyncFunction("runCommand") { (command: String, arguments: [String]) -> String in
      let task = Process()
      let pipe = Pipe()
      
      task.executableURL = URL(string: "/bin/sh")
      task.arguments = ["-l", "-c", "PATH=\"$PATH:/usr/local/bin\"; \(command) \(arguments.joined(separator: " "))"]
      task.standardOutput = pipe
      task.standardError = pipe
      
      let file = pipe.fileHandleForReading
      var returnOutput = ""
      
      try task.run()
      
      let notificationCenter = NotificationCenter.default
      
      notificationCenter.addObserver(forName: FileHandle.readCompletionNotification, object: file, queue: nil) { [weak self] notifiction in
        guard let self else {
          return
        }
        guard let chunk = notifiction.userInfo?[NSFileHandleNotificationDataItem] as? Data else {
          return
        }
        let wholeOutput = String(data: chunk, encoding: .utf8)
        guard let outputs = wholeOutput?.components(separatedBy: .newlines) else {
          return
        }
        
        for output in outputs {
          if output.isEmpty {
            continue
          }
          
          returnOutput = returnOutput.appending(output)
          if hasListeners && !output.isEmpty && output != "\n" {
            sendEvent(onNewCommandLine,
                    ["output": output]
            )
          }
        }
        
        file.readInBackgroundAndNotify()
      }
      
      notificationCenter.addObserver(forName: Process.didTerminateNotification, object: task, queue: nil) { notification in
        notificationCenter.removeObserver(self)
      }
      
      file.readInBackgroundAndNotify()
      task.waitUntilExit()
      return returnOutput
    }
    
    AsyncFunction("runCli") { (command: String, arguments: [String], listenerId: Int, promise: Promise) in
      let task = Process()
      let pipe = Pipe()
      
      task.executableURL = Bundle.main.url(forResource: getCliResourceNameForArch(), withExtension: nil)
      task.arguments = [command] + arguments
      
      var environment = ProcessInfo.processInfo.environment
      environment["EXPO_MENU_BAR"] = "true"
      
      if let envVars = UserDefaults.standard.object(forKey: "envVars") as? Dictionary<String, String> {
        if envVars.count > 0 {
          environment = environment.merging(envVars) { _, new in
            new
          }
        }
      }
      
      task.environment = environment
      task.standardOutput = pipe
      task.standardError = pipe
      
      let file = pipe.fileHandleForReading
      var returnOutput = ""
      var hasReachedReturnOutput = false
      var hasReachedError = false
      
      file.readabilityHandler = { [weak self] handle in
        guard let self else {
          return
        }
        
        let availableData = handle.availableData
        let wholeOutput = String(data: availableData, encoding: .utf8)
        guard let outputs = wholeOutput?.components(separatedBy: .newlines) else {
          return
        }
        
        for output in outputs {
          if output.isEmpty {
            continue
          }
          
          if hasReachedReturnOutput || hasReachedError {
            returnOutput.append(output)
          } else if output == "---- return output ----" {
            hasReachedReturnOutput = true
          } else if output == "---- thrown error ----" {
            hasReachedError = true
          } else if hasListeners && !output.isEmpty && output != "\n" {
            let eventData = ["listenerId": listenerId, "output": output]
            sendEvent(onCLIOutput, eventData)
          }
        }
      }
      
      task.terminationHandler = { task in
        file.readabilityHandler = nil
        
        if hasReachedError {
          promise.reject(CLIOutputError(returnOutput))
        } else {
          if task.terminationStatus == 0 {
            promise.resolve(hasReachedReturnOutput ? returnOutput : nil)
          } else {
            promise.reject(IntenalCLIError(fullOutput))
          }
        }
      }
      
      try task.run()
    }
    
    AsyncFunction("setLoginItemEnabled") { (enabled: Bool) in
      guard SMLoginItemSetEnabled("dev.expo.AutoLauncher" as CFString, enabled) else {
        throw LauncherError()
      }
    }
    
    AsyncFunction("showMultiOptionAlert") { (title: String, message: String, options: [String]) -> Int in
      let alert = NSAlert()
      alert.messageText = title
      alert.informativeText = message
      alert.addButton(withTitle: "OK")
      alert.addButton(withTitle: "Cancel")
      
      let popupButton = NSPopUpButton(frame: NSRect(x: 0, y: 0, width: 250, height: 24), pullsDown: false)
      popupButton.addItems(withTitles: options)
      alert.accessoryView = popupButton
      
      if alert.runModal() == .alertFirstButtonReturn {
        return popupButton.indexOfSelectedItem
      }
      throw MultiOptionError()
    }.runOnQueue(.main)
    
    AsyncFunction("openSystemSettingsLoginItems") {
      if #available(macOS 13.0, *) {
        SMAppService.openSystemSettingsLoginItems()
      } else {
        if let url = URL(string: "x-apple.systempreferences:com.apple.LoginItems-Settings.extension") {
          NSWorkspace.shared.open(url)
        }
      }
    }
    
    AsyncFunction("setEnvVars") { (envVars: [String: String]) in
      UserDefaults.standard.setValue(envVars, forKey: "envVars")
    }
    
    // Run on the main queue because NotificationCenter needs to post from the same queue it will receive on
    AsyncFunction("openPopover") {
      NotificationCenter.default.post(name: NSNotification.Name(rawValue: "ExpoOrbit_OpenPopover"), object: nil)
    }.runOnQueue(.main)
    
    AsyncFunction("closePopover") {
      NotificationCenter.default.post(name: NSNotification.Name(rawValue: "ExpoOrbit_ClosePopover"), object: nil)
    }.runOnQueue(.main)
    
    OnStartObserving {
      hasListeners = true
    }
    
    OnStopObserving {
      hasListeners = false
    }
  }
  
  private func getHardwareArch() -> String? {
    var sysinfo = utsname()
    let result = uname(&sysinfo)
    
    if result != 0 {
      return nil
    }
    
    let machine = withUnsafePointer(to: &sysinfo.machine) {
      $0.withMemoryRebound(to: CChar.self, capacity: Int(_SYS_NAMELEN)) {
        ptr in String(cString: ptr)
      }
    }
    
    return machine
  }
  
  private func getCliResourceNameForArch() -> String {
    return getHardwareArch() == "arm64" ? "orbit-cli-arm64" : "orbit-cli-x64"
  }
}
